lib/internal/Magento/Framework/App/Http.php

```diff
@@ -142,12 +142,26 @@ public function launch()
         } else {
             throw new \InvalidArgumentException('Invalid return type');
         }
+        if ($this->_request->isHead() && $this->_response->getHttpResponseCode() == 200) {
+            $this->handleHeadRequest();
+        }
         // This event gives possibility to launch something before sending output (allow cookie setting)
         $eventParams = ['request' => $this->_request, 'response' => $this->_response];
         $this->_eventManager->dispatch('controller_front_send_response_before', $eventParams);
         return $this->_response;
     }
 
+    /**
+     * Handle HEAD requests by adding the Content-Length header and removing the body from the response.
+     * @return void
+     */
+    private function handleHeadRequest()
+    {
+        $contentLength = strlen($this->_response->getContent());
```

Comment:
The strlen function isn’t binary safe (i.e. images and compressed content) and also doesn’t handle multi-byte characters. I‘m not sure about the proper things to do here, but it would  be a good idea to do some more research and maybe add some tests for non-ascii response bodies.

Reply:
I was under the assumption that strlen specifically returns the number of bytes in the string, instead of the number of characters.[1] Isn't that what the Content-Length header should contain?

[1] http://php.net/manual/en/function.strlen.php#refsect1-function.strlen-notes

Reply:
I do agree on adding non-ascii test cases, that is a good idea.

Reply:
On further research it seems that some installations of PHP might have overloaded the strlen to use mb_strlen instead.[1] This would break the content length on multibyte characters as they would be counted as one, instead of the actual number of bytes.

Safer way seems to be to call mb_strlen explicitly with encoding that only supports byte characters.[2] This way we should always get the correct byte count for the body contents.

[1] http://php.net/manual/en/mbstring.overload.php
[2] https://stackoverflow.com/a/9718273

Reply:
I guess the important thing to check for would be if it is null byte safe `\0`. That might be PHP version dependent, hopefully PHP7 does not rely on low level C strings. Otherwise your suggested use of `mb_strlen` sounds good to me.

Reply:
As far as I can see, the null byte is treated as a one byte character. It won't terminate strings prematurely.[1]

[1] https://stackoverflow.com/a/32263348

Reply:
Updated the pull request to use mb_strlen instead of strlen and added test cases for some potentially anomalous content.